### PR TITLE
Add DataManager to AuxCoord

### PR DESCRIPTION
**Pull request by djkirkham**

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -1753,13 +1757,7 @@ def _sanitise_array(self, src, ndmin):
...
     def points(self):
         """Property containing the points values as a numpy array"""
-        if is_lazy_data(self._points):
-            self._points = as_concrete_data(self._points,
-                                            nans_replacement=np.ma.masked)
-            # NOTE: we probably don't have full support for masked aux-coords.
-            # We certainly *don't* handle a _FillValue attribute (and possibly
-            # the loader will throw one away ?)
-        return self._points.view()
+        return self._points_dm.data.view()
```
Comment (dkillick) [inline lib/iris/coords.py]: I think we want to lose the `view()` call now don't we?

File: lib/iris/coords.py [hunk 2]
```diff
@@ -1792,20 +1789,10 @@ def bounds(self):
...
-                # possibly the loader will throw one away ?)
-                self._bounds = bounds
-            bounds = bounds.view()
+        if self._bounds_dm is not None:
+            return self._bounds_dm.data.view()
         else:
-            bounds = None
-
-        return bounds
+            return None
```
Comment (dkillick) [inline lib/iris/coords.py]: I prefer the [approach that I've taken](https://github.com/dkillick/iris/blob/b9be68bb4174f609b0b7a28ad05ec18e8b915315/lib/iris/coords.py#L1691-L1701) that does not require multiple return statements. Multiple return statements can make code hard to follow and debug.

File: lib/iris/coords.py [hunk 3]
```diff
@@ -1814,10 +1801,15 @@ def bounds(self, bounds):
             if not is_lazy_data(bounds):
                 bounds = self._sanitise_array(bounds, 2)
             # NB. Use _points to avoid triggering any lazy array.
-            if self._points.shape != bounds.shape[:-1]:
+            if self._points_dm.shape != bounds.shape[:-1]:
```
Comment (dkillick) [inline lib/iris/coords.py]: `self.shape`, surely?

File: lib/iris/coords.py [hunk 4]
```diff
@@ -1814,10 +1801,15 @@ def bounds(self, bounds):
...
                 raise ValueError("Bounds shape must be compatible with points "
                                  "shape.")
-        self._bounds = bounds
+            if self._bounds_dm is None:
+                self._bounds_dm = DataManager(bounds)
```
Comment (dkillick) [inline lib/iris/coords.py]: You might want to specify a `dtype` for this too, if possible.

File: lib/iris/coords.py (hunk 1 above)
Comment (djkirkham) [inline lib/iris/coords.py]: I don't know. I read some comments from @pp-mo saying we shouldn't be returning views, but why not? Don't we want to prevent users altering the underlying data array? Even if the array is not writeable, they can still alter meta-data like shape and dtype.

File: lib/iris/coords.py (hunk 2 above)
Comment (djkirkham) [inline lib/iris/coords.py]: Personally I find this easier to read, but I'll change it if you like.

File: lib/iris/coords.py (hunk 3 above)
Comment (djkirkham) [inline lib/iris/coords.py]: ok

File: lib/iris/coords.py (hunk 4 above)
Comment (djkirkham) [inline lib/iris/coords.py]: I the bounds are being set through the `bounds` setter then I think we have to assume that the dtype is to be the same as the underlying array, which is handled by passing no dtype to `DataManager.__init__`. Otherwise `replace()` should be used.

**Comment by dkillick** [comment]
I'm going to merge this as-is into the main PR (https://github.com/SciTools/iris/pull/2527). We can address outstanding issues and API changes after the merge.
